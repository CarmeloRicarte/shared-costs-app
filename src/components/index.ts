export * from './CustomTable/CustomTable';
export * from './Modal';
