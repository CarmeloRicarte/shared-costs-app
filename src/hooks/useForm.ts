import { ChangeEvent, useEffect, useState } from 'react';

export const useForm = <T extends object>(initialForm: T) => {
  const [formState, setFormState] = useState(initialForm);

  useEffect(() => {
    setFormState(initialForm);
  }, []);

  /**
   * When the input changes, update the state of the form with the new value of the input.
   */
  const onInputChange = ({
    target,
  }: ChangeEvent<
    HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement
  >) => {
    const { name, value, type } = target;
    const newValue =
      type === 'number' && value !== '' ? parseFloat(value) : value;
    setFormState({
      ...formState, // maintain state of other form fields
      [name]: newValue, // update state of field that is changed
    });
  };

  /**
   * It resets the form to its initial state.
   */
  const onResetForm = () => {
    setFormState(initialForm);
  };

  return {
    ...formState, // for return every form control of state
    formState,
    onInputChange,
    onResetForm,
  };
};
