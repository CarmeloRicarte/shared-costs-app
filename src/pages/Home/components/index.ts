export * from './Balance';
export * from './CostsList';
export * from './CostsListItem';
