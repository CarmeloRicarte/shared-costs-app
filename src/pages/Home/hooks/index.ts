export * from './useBalances';
export * from './useCosts';
export * from './useFriends';
export * from './useGroups';
