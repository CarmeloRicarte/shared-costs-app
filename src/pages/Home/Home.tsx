import { useEffect, useState } from 'react';

import { Modal } from '../../components';
import { LoadingSpinner } from '../../components/LoadingSpinner';
import { AddFriendForm, AddPaymentForm, BalanceByFriend } from './components';
import { CostsList } from './components/CostsList';
import { useCosts } from './hooks';

export const Home = () => {
  const [isModalAddFriendOpen, setIsModalAddFriendOpen] = useState(false);
  const [isModalAddPaymentOpen, setIsModalAddPaymentOpen] = useState(false);
  const { costs, getAllCosts, isLoading } = useCosts();

  const handleToggleAddFriendModal = () => {
    setIsModalAddFriendOpen(!isModalAddFriendOpen);
  };

  const handleToggleAddPaymentModal = () => {
    setIsModalAddPaymentOpen(!isModalAddPaymentOpen);
  };

  const getCostsData = async () => {
    try {
      await getAllCosts();
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    void getCostsData();
  }, []);

  return (
    <>
      <h1 className='text-center'>Gastos Compartidos</h1>
      <div className='list-grid-2-columns'>
        {isLoading ? (
          <LoadingSpinner />
        ) : (
          <>
            <CostsList
              costs={costs}
              onClickPagoButton={handleToggleAddPaymentModal}
              onClickAmigoButton={handleToggleAddFriendModal}
            />
            <BalanceByFriend />
          </>
        )}
      </div>
      <Modal
        isOpen={isModalAddFriendOpen}
        onClose={handleToggleAddFriendModal}
        headerTitle='Añadir amigo al grupo'
      >
        <AddFriendForm onSubmitForm={handleToggleAddFriendModal} />
      </Modal>
      <Modal
        isOpen={isModalAddPaymentOpen}
        onClose={handleToggleAddPaymentModal}
        headerTitle='Añadir pago'
      >
        <AddPaymentForm onSubmitForm={handleToggleAddPaymentModal} />
      </Modal>
    </>
  );
};
