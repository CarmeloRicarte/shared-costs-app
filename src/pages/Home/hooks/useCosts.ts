import { useState } from 'react';

import { useLocalStorage } from '../../../hooks';
import { useHomeContext } from '../context';
import { Cost } from '../models';
import { getCosts } from '../services';

export const useCosts = () => {
  const { costs, setCosts } = useHomeContext();
  const { setItem, getItem } = useLocalStorage();
  const [isLoading, setisLoading] = useState(false);

  const setCostsContextAndLocalStorage = (costs: Cost[]) => {
    setCosts(costs);
    setItem('costs', costs);
  };

  const getAllCosts = async () => {
    setisLoading(true);
    if (getItem('costs')) {
      setCostsContextAndLocalStorage(getItem('costs'));
    } else {
      const costs = await getCosts();
      costs && setCostsContextAndLocalStorage(costs);
    }
    setisLoading(false);
  };

  const createCost = (cost: Cost) => {
    const newCosts = [...costs, cost];
    setCostsContextAndLocalStorage(newCosts);
  };

  return {
    costs,
    getAllCosts,
    createCost,
    setCostsContextAndLocalStorage,
    isLoading,
  };
};
