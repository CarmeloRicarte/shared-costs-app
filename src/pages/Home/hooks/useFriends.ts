import { useEffect } from 'react';

import { useLocalStorage } from '../../../hooks';
import { useHomeContext } from '../context';
import { Friend } from '../models';
import { createFriend, getFriends } from '../services';

export const useFriends = () => {
  const { friends, setFriends } = useHomeContext();
  const { setItem, getItem } = useLocalStorage();

  useEffect(() => {
    void getAllFriends();
  }, []);

  const setFriendsContextAndLocalStorage = (friends: Friend[]) => {
    setFriends(friends);
    setItem('friends', friends);
  };

  const getAllFriends = async () => {
    if (getItem('friends')) {
      setFriendsContextAndLocalStorage(getItem('friends'));
    } else {
      const friends = await getFriends();
      friends && setFriendsContextAndLocalStorage(friends);
    }
  };

  const addFriend = async (friend: Friend) => {
    const createdFriend = await createFriend(friend);
    if (createdFriend) {
      const newFriends = [...friends, friend];
      setFriendsContextAndLocalStorage(newFriends);
    }
  };

  return {
    friends,
    getAllFriends,
    addFriend,
  };
};
