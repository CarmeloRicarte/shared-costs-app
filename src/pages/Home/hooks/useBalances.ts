import { useEffect } from 'react';

import { useHomeContext } from '../context';
import { Balance } from '../models';
import { useCosts } from './useCosts';

export const useBalances = () => {
  const { costs } = useCosts();
  const { balances, setBalances } = useHomeContext();
  useEffect(() => {
    costs.length > 0 && calculateBalance();
  }, [costs, balances]);

  /**
   * It takes an array of objects, and returns an array of objects, where each object has a friendName
   * and a balance
   * @returns An array of objects with the friendName and balance properties.
   */
  const getCostsPaid = () => {
    const costsMapped = costs.map((cost) => ({
      friendName: cost.personName,
      balance: cost.totalAmount,
    }));

    return costsMapped.reduce((acc: Balance[], { friendName, balance }) => {
      const personExisting = acc.find(
        (personBalance) => personBalance.friendName === friendName
      );

      if (personExisting !== undefined) {
        personExisting.balance += balance;
      } else {
        acc.push({ friendName, balance });
      }

      return acc;
    }, []);
  };

  /**
   * It calculates the balance of each friend by subtracting the average amount paid by each friend
   * from the total amount paid by each friend
   */
  const calculateBalance = () => {
    const costsPaidByFriend = getCostsPaid();
    const totalPaid = costsPaidByFriend.reduce(
      (acc, person) => acc + person.balance,
      0
    );
    const average = totalPaid / costsPaidByFriend.length;
    const balances = costsPaidByFriend.map((friendBalance) => ({
      friendName: friendBalance.friendName,
      balance: friendBalance.balance - average,
    }));
    setBalances(balances);
  };

  return {
    balances,
  };
};
