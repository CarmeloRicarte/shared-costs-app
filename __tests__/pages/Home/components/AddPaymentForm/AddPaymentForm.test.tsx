import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { AddPaymentForm } from 'pages/Home/components';
import * as useBalances from 'pages/Home/hooks/useBalances';
import * as useCosts from 'pages/Home/hooks/useCosts';
import * as useFriends from 'pages/Home/hooks/useFriends';

import { mockFriends } from '../../../../__fixtures__';

vi.mock('pages/Home/hooks/useCosts');
vi.mock('pages/Home/hooks/useFriends');
vi.mock('pages/Home/hooks/useBalances');

describe('AddFriendForm tests', () => {
  const onSubmitFormMock = vi.fn();
  const createCostMock = vi.fn();
  const calculateBalanceMock = vi.fn();

  beforeEach(() => {
    (useCosts as any).useCosts = vi.fn().mockReturnValue({
      createCost: createCostMock,
    });

    (useFriends as any).useFriends = vi.fn().mockReturnValue({
      friends: mockFriends,
    });

    (useBalances as any).useBalances = vi.fn().mockReturnValue({
      calculateBalance: calculateBalanceMock,
    });
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  test('should render component', () => {
    const { container } = render(
      <AddPaymentForm onSubmitForm={onSubmitFormMock} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should call to createCost onSubmit form with a valid payment', async () => {
    render(<AddPaymentForm onSubmitForm={onSubmitFormMock} />);
    const personNameSelect = screen.getByTestId('personName');
    const descriptionInput = screen.getByTestId('description');
    const totalAmountInput = screen.getByTestId('totalAmount');
    const paymentDateInput = screen.getByTestId('paymentDate');

    const submitButton = screen.getByRole('button', { name: 'AÃ±adir' });
    fireEvent.change(personNameSelect, { target: { value: 'Pedro' } });
    fireEvent.change(descriptionInput, { target: { value: 'test' } });
    fireEvent.change(totalAmountInput, { target: { value: '4.50' } });
    fireEvent.change(paymentDateInput, {
      target: { value: '2023-02-18T22:00:00' },
    });
    fireEvent.submit(submitButton);

    await waitFor(() => {
      expect(createCostMock).toHaveBeenCalledWith({
        id: expect.any(String),
        personName: 'Pedro',
        description: 'test',
        totalAmount: 4.5,
        paymentDate: '2023-02-18T22:00',
      });
    });
    expect(calculateBalanceMock).toHaveBeenCalled();
    expect(onSubmitFormMock).toHaveBeenCalled();
  });
});
