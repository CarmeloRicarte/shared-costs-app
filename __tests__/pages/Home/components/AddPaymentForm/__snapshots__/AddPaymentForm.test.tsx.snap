// Vitest Snapshot v1

exports[`AddFriendForm tests > should render component 1`] = `
<div>
  <form
    class="payment-group-form"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-2"
        >
          <label
            class="payment-group-form-label"
            for="personName"
          >
            Nombre persona:
          </label>
        </div>
        <div
          class="col-4"
        >
          <select
            class="payment-group-form-select"
            data-testid="personName"
            id="personName"
            name="personName"
            required=""
          >
            <option
              value=""
            >
              Seleccione
            </option>
            <option
              value="Pedro"
            >
              Pedro
            </option>
          </select>
        </div>
        <div
          class="col-2"
        >
          <label
            class="payment-group-form-label"
            for="description"
          >
            Descripción:
          </label>
        </div>
        <div
          class="col-4"
        >
          <textarea
            class="payment-group-form-input"
            data-testid="description"
            id="description"
            name="description"
            required=""
          />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-2"
        >
          <label
            class="payment-group-form-label"
            for="totalAmount"
          >
            Importe (€):
          </label>
        </div>
        <div
          class="col-4"
        >
          <input
            class="payment-group-form-input"
            data-testid="totalAmount"
            id="totalAmount"
            min="0.01"
            name="totalAmount"
            placeholder="0,00 €"
            required=""
            step="any"
            type="number"
            value="0"
          />
        </div>
        <div
          class="col-2"
        >
          <label
            class="payment-group-form-label"
            for="paymentDate"
          >
            Fecha pago:
          </label>
        </div>
        <div
          class="col-4"
        >
          <input
            class="payment-group-form-input"
            data-testid="paymentDate"
            id="paymentDate"
            name="paymentDate"
            required=""
            type="datetime-local"
            value=""
          />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-1 col-offset-11 text-end"
        >
          <button
            class="payment-group-form-button"
            disabled=""
            id="addFriendButton"
            type="submit"
          >
            Añadir
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
